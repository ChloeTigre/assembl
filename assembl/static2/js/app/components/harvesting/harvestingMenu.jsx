// @flow
import * as React from 'react';

import addPostExtractMutation from '../../graphql/mutations/addPostExtract.graphql'; // eslint-disable-line
import updateExtractMutation from '../../graphql/mutations/updateExtract.graphql'; // eslint-disable-line
import deleteExtractMutation from '../../graphql/mutations/deleteExtract.graphql'; // eslint-disable-line
import HarvestingAnchor from './harvestingAnchor';
import HarvestingBox from './harvestingBox';
import HarvestingBadge from './harvestingBadge';
import { getAnnotationData } from '../../utils/extract';

type Props = {
  extracts: Array<Extract>,
  postId: string,
  lang: string,
  displayHarvestingBox: boolean,
  displayHarvestingAnchor: boolean,
  harvestingAnchorPosition: Object,
  setHarvestingBoxDisplay: Function,
  handleClickAnchor: Function,
  cancelHarvesting: Function,
  refetchPost: Function,
  isAuthorAccountDeleted: boolean,
  showNuggetAction: boolean
};

type State = {
  displayExtractsBox: boolean
};

class HarvestingMenu extends React.Component<Props, State> {
  static getDerivedStateFromProps(props: Props) {
    const { extracts } = props;
    const { hash } = window.location;
    if (hash !== '') {
      const hashExtractId = hash.split('#')[2];
      let isHashMatchId = false;
      extracts.forEach((extract) => {
        if (hashExtractId === extract.id) isHashMatchId = true;
      });
      return { displayExtractsBox: isHashMatchId };
    }
    return null;
  }

  state = { displayExtractsBox: false };

  toggleExtractsBox = (): void => {
    this.setState(prevState => ({ displayExtractsBox: !prevState.displayExtractsBox }));
  };

  handleMouseDown = (event: SyntheticMouseEvent<>) => {
    // This would otherwise clear the selection
    event.preventDefault();
    return false;
  };

  render() {
    const {
      setHarvestingBoxDisplay,
      displayHarvestingBox,
      displayHarvestingAnchor,
      postId,
      handleClickAnchor,
      extracts,
      harvestingAnchorPosition,
      refetchPost,
      cancelHarvesting,
      isAuthorAccountDeleted,
      showNuggetAction,
      lang
    } = this.props;
    const selection = window.getSelection();
    const annotation = getAnnotationData(selection);
    const hasAnnotation = !!annotation;
    const { displayExtractsBox } = this.state;
    const showHarvestingBadge = !displayExtractsBox && extracts && extracts.length > 0;
    const showBoxWithExtracts = displayExtractsBox && extracts && extracts.length > 0 && !displayHarvestingBox;
    const showBoxInHarvestingMode = displayHarvestingBox && hasAnnotation;
    const showHarvestingAnchor = displayHarvestingAnchor && hasAnnotation;
    return (
      <div>
        {showHarvestingBadge && <HarvestingBadge toggleExtractsBox={this.toggleExtractsBox} extractsNumber={extracts.length} />}
        <div className="harvesting-container">
          {showBoxWithExtracts && (
            <HarvestingBox
              postId={postId}
              key={`extracts-${postId}`}
              extracts={extracts}
              isAuthorAccountDeleted={isAuthorAccountDeleted}
              displayHarvestingBox={displayHarvestingBox}
              refetchPost={refetchPost}
              cancelHarvesting={cancelHarvesting}
              setHarvestingBoxDisplay={setHarvestingBoxDisplay}
              showNuggetAction={showNuggetAction}
              toggleExtractsBox={this.toggleExtractsBox}
            />
          )}
          {showBoxInHarvestingMode && (
            <HarvestingBox
              postId={postId}
              key={`harvesting-${postId}`}
              annotation={annotation}
              lang={lang}
              displayHarvestingBox={displayHarvestingBox}
              refetchPost={refetchPost}
              cancelHarvesting={cancelHarvesting}
              setHarvestingBoxDisplay={setHarvestingBoxDisplay}
              showNuggetAction={showNuggetAction}
            />
          )}
          {showHarvestingAnchor && (
            <HarvestingAnchor
              displayHarvestingBox={displayHarvestingBox}
              handleMouseDown={this.handleMouseDown}
              anchorPosition={harvestingAnchorPosition}
              handleClickAnchor={handleClickAnchor}
            />
          )}
        </div>
      </div>
    );
  }
}

export default HarvestingMenu;