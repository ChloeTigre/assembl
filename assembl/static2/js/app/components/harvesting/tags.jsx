// @flow
import React from 'react';
import { I18n } from 'react-redux-i18n';
import { Button } from 'react-bootstrap';

import TagsForm, { type FormData } from './tagsForm';

type Props = {
  initialValues: Array<string>,
  canEdit: boolean,
  updateTags: (tags: Array<string>, callback: (tags: Array<string>) => void) => void
};

type State = {
  editing: false,
  submitting: boolean,
  currentTags: Array<string>
};

export type TagsData = {
  tags: Array<string>
};

class Tags extends React.Component<Props, State> {
  static defaultProps = {
    initialValues: [],
    canEdit: false
  };

  state = {
    editing: false,
    submitting: false,
    currentTags: this.props.initialValues
  };

  renderFooter = ({ handleSubmit, pristine = false, submitting = false }: FormData) => (
    <div className="harvesting-box-footer">
      <Button disabled={submitting} key="cancel" className="button-cancel button-dark" onClick={this.cancel}>
        {I18n.t('harvesting.tags.cancel')}
      </Button>
      <Button disabled={pristine || submitting} key="validate" className="button-submit button-dark" onClick={handleSubmit}>
        {I18n.t('harvesting.tags.validate')}
      </Button>
    </div>
  );

  updateTags = (data: TagsData) => {
    const { updateTags } = this.props;
    const { tags } = data;
    this.setState(
      {
        submitting: true
      },
      () => {
        updateTags(tags, (newtags: Arra<string>) => {
          this.setState({
            submitting: false,
            editing: false,
            currentTags: newtags
          });
        });
      }
    );
  };

  cancel = () => {
    const { initialValues } = this.props;
    this.setState({
      editing: false,
      currentTags: initialValues
    });
  };

  edit = () => {
    this.setState({
      editing: true
    });
  };

  pristine = (currentTags: Array<string>) => {
    const { initialValues } = this.props;
    return currentTags.length === initialValues.length && currentTags.every(tag => initialValues.includes(tag));
  };

  removeTag = (tag: string) => {
    this.setState(state => ({
      ...state,
      currentTags: state.currentTags.filter(currentTag => currentTag !== tag)
    }));
  };

  renderTags = () => {
    const { canEdit } = this.props;
    const { currentTags, submitting } = this.state;
    const pristine = this.pristine(currentTags);
    return (
      <div className="harvesting-tags-container">
        <label htmlFor="tags">{I18n.t('harvesting.tags.label')}</label>
        <div className="harvesting-tags-titles-container">
          {currentTags.map(tag => (
            <div key={tag} className="harvesting-tag-container">
              <span className="harvesting-tag-title">{tag}</span>
              {canEdit ? <div className="assembl-icon-cancel" onClick={() => this.removeTag(tag)} /> : null}
            </div>
          ))}
          {canEdit ? (
            <div className="harvesting-tags-edit" onClick={this.edit}>
              +
            </div>
          ) : null}
        </div>
        {!pristine &&
          this.renderFooter({
            handleSubmit: () => this.updateTags({ tags: currentTags }),
            pristine: pristine,
            submitting: submitting
          })}
      </div>
    );
  };

  renderForm = () => {
    const { currentTags } = this.state;
    return <TagsForm initialValues={currentTags} onSubmit={this.updateTags} renderFooter={this.renderFooter} />;
  };

  render() {
    const { canEdit } = this.props;
    const { editing } = this.state;
    return canEdit && editing ? this.renderForm() : this.renderTags();
  }
}

export default Tags;