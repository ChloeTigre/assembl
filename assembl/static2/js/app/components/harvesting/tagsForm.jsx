// @flow
import React from 'react';
import { Translate, I18n } from 'react-redux-i18n';
import { Form, Field } from 'react-final-form';
import { type ApolloClient, withApollo } from 'react-apollo';

import SelectFieldAdapter from '../form/selectFieldAdapter';
import TagsQuery from '../../graphql/TagsQuery.graphql';

export type FormData = {
  handleSubmit: Function,
  pristine: boolean,
  submitting: boolean
};

type Props = {
  client: ApolloClient,
  initialValues: Array<string>,
  renderFooter: (formData: FormData) => React.Node,
  onSubmit: ({ tags: Array<string> }) => void
};

export const tagsLoader = (client: ApolloClient) => inputValue =>
  client
    .query({
      query: TagsQuery,
      variables: { filter: inputValue }
    })
    .then((value) => {
      const { data: { tags } } = value;
      return tags.map(tag => ({
        label: tag.value,
        value: tag.value
      }));
    });

class TagsForm extends React.Component<Props> {
  static defaultProps = {
    initialValues: []
  };

  getTagsLoader = () => {
    const { client } = this.props;
    return tagsLoader(client);
  };

  pristine = (currentTags) => {
    const { initialValues } = this.props;
    return currentTags.length === initialValues.length && currentTags.every(tag => initialValues.includes(tag));
  };

  render() {
    const { initialValues, renderFooter, onSubmit } = this.props;
    return (
      <Form
        initialValues={{ tags: initialValues }}
        onSubmit={onSubmit}
        render={({ handleSubmit, values, submitting }) => {
          const pristine = this.pristine(values.tags);
          return (
            <form onSubmit={handleSubmit} className="harvesting-tags-form-container form-container">
              <Field
                name="tags"
                component={SelectFieldAdapter}
                label={I18n.t('harvesting.tags.label')}
                // The Select fields
                isMulti
                canCreate
                isAsync
                value={initialValues}
                loadOptions={this.getTagsLoader()}
                className="tags-select"
                placeholder="harvesting.tags.select.placeholder"
                noOptionsMessage={() => <Translate value="harvesting.tags.select.noOptions" />}
                formatCreateLabel={newOption => <Translate value="harvesting.tags.select.newOption" option={newOption} />}
              />
              {renderFooter({ handleSubmit: handleSubmit, pristine: pristine, submitting: submitting })}
            </form>
          );
        }}
      />
    );
  }
}

export default withApollo(TagsForm);