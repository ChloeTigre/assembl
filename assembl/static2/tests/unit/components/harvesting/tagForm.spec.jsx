import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbTagForm } from '../../../../js/app/components/harvesting/tagForm';

describe('Tag component', () => {
  it('should match the Tag form', () => {
    const props = {
      initialValue: { label: 'foo', value: 'fooid' },
      onCancel: jest.fn(() => {}),
      onSubmit: jest.fn(() => {})
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbTagForm {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});