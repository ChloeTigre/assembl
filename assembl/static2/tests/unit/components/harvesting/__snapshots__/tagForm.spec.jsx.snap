// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag component should match the Tag form 1`] = `
<ReactFinalForm
  initialValues={
    Object {
      "tag": Object {
        "label": "foo",
        "value": "fooid",
      },
    }
  }
  onSubmit={[MockFunction]}
  render={[Function]}
/>
`;
